src/select.rs

```diff
@@ -1,66 +1,149 @@
-use crate::compare::*;
-use crate::utils::*;
-use crate::NNFitnessFunction;
+use crate::generation::Generation;
+use crate::utils::Individual;
+use crate::FitnessFunction;
+use cge::gene::Gene;
+use cge::gene::GeneExtras::*;
 
 // For preserving diversity
 const MAX_DUPLICATES: usize = 1;
 const MAX_SIMILAR: usize = 2;
 
-pub fn select<T>(population_size: usize, individuals: Vec<Individual<T>>, threshold: f64) -> Vec<Individual<T>>
-    where T: NNFitnessFunction + Clone
-{
-    let mut pre_generation = Vec::new();
+/// 1. Greedily remove as many (purely structural) duplicates / too-similar networks as possible
+/// 2. Take the remaining individuals who are most fit
+// TODO: this is `O(n^2)`, optimization may be possible.
+// TODO: I am thinking that if we can figure out a workable network "fingerprint", this could run in `O(n)`,
+// TODO: and other speedups could be achieved.
+// this is called serially in the main EANT2 algorithm,
+// so (for now) there would be some benefit to parallelizing it. I got stuck doing this.
```

Comment:
If each network must be compared to each other, I'm not sure if it would be possible to reduce it past `O(n^2)`. It could at least avoid redundant comparisons by starting the second loop from the current index. However, this entire step may end up being only a tiny fraction of the overall running time, making any such optimizations negligible.

Reply:
Yes, probably not worth the effort!

I think strictly speaking it could be done in less than `O(n^2)`, but it would require a very significant change to how it works.  In general, a "clustering" cannot be formed, but every network need not be compared to all other networks for structural deduplication.  A fingerprint (hash) could be resolved in O(1) for each item (check if we've already seen it), so exact duplicates could be eliminated in O(n), and I think other tricks could be played for O(n log n) complexity on the remaining elements.  Perhaps a formalization would clear up /how/ to do this, e.g. questions like the associativity of similarity.  Maybe if I'm bored someday, or if it turns out to actually matter.